runtime/lua/vim/keymap.lua

```diff
@@ -0,0 +1,136 @@
+--- Examples:
+--- <pre>
+---   -- You can make keymaps with Lua functions
+---   vim.keymap.set('n', 'lhs', function() print("real lua function") end)
+---
+---   -- Can use it to map multiple modes
+---   vim.keymap.set({'n', 'v'}, '<leader>lr', vim.lsp.buf.references, { silent = true, buffer=true })
+---
+---   -- You can map a key for specific buffer
+---   vim.keymap.set('n', '<leader>w', "<cmd>w<cr>", { silent = true, buffer = 5 })
+---
+---   -- Also expr keymaps too. Notice the replace_termcodes when making Lua expr maps.
+---   vim.keymap.set('i', '<Tab>', function()
+---     return vim.fn.pumvisible() == 1 and "<C-n>" or "<Tab>"
+---   end, { silent = true, expr = true, replace_termcodes = true })
+---   -- While setting <Plug> mapping remember to set remap to true
+---   vim.keymap.set('n', '[%', '<Plug>(MatchitNormalMultiBackward)', {silent = true, remap = true })
+--- </pre>
+---
+--- A pit fall of this is if you want to make keymap like:
+--- <pre>
+---    vim.keymap.set('n', 'asdf', require('jkl').my_fun)
+--- </pre>
+---
+--- `require('jkl)` get's evaluated durimg this call. If you want to
+--- avoid this cost at startup you can wrap it in a function, for example:
+--- <pre>
+---    vim.keymap.set('n', 'asdf', function() return require('jkl').my_fun() end)
+--- </pre>
+---
+---@param mode string|table   Same mode short names as |nvim_set_keymap()|.
+---                            Can also be list of modes to create keymap on multiple modes.
+---@param lhs string          Left-hand-side |{lhs}| of the mapping.
+---@param rhs string|function  Right-hand-side |{rhs}| of the mapping. Can also be a Lua function.
+--
+---@param opts table A table of |:map-arguments| such as "silent". In addition to the options
+---                  listed in |nvim_set_keymap()|, this table also accepts the following keys:
+---                  - replace_termcodes: (boolean, default false) when "true",
+---                  |nvim_replace_termcodes()| is applied to the result of Lua expr maps.
+---                  - remap: (boolean, default false) Make the mapping recursive. This is the
+---                  inverse of the "noremap" option from |nvim_set_keymap()|.
+---@see |nvim_set_keymap()|
+function keymap.set(mode, lhs, rhs, opts)
+  vim.validate {
+    mode = {mode, {'s', 't'}},
+    lhs = {lhs, 's'},
+    rhs = {rhs, {'s', 'f'}},
+    opts = {opts, 't', true}
+  }
+
+  opts = opts or {}
+  local buffer = false
+  local is_rhs_lua = type(rhs) == "function"
+  mode = type(mode) == 'string' and {mode} or mode
+
+  if opts.buffer ~= nil then
+    buffer = opts.buffer == true and 0 or opts.buffer
+    opts.buffer = nil
+  end
+
+  if opts.replace_termcodes ~= nil then
+    if opts.replace_termcodes and is_rhs_lua then
+      local user_rhs = rhs
+      rhs = function ()
+        return vim.api.nvim_replace_termcodes(user_rhs(), true, true, true)
+      end
+    end
+    opts.replace_termcodes = nil
+  end
+
+  -- remap defaults to false
```

Comment:
is it crazy to check something like `<Plug>` in the rhs if it's a string and then turn remap on by default?

Reply:
If it's not considered too magical for core, I have no issue with it 😆 . 

Should we be changing default value of an option based on other option ? I can definitely see this being more convenient . The whole remap noremap thing is a mess for any new comer .

Reply:
Could also leave it for a separate follow-up PR? I don't think adding that would be a breaking change (as opposed to removing it).

Reply:
It's 3 lines of code we can just decide whether we want it or not . `<Plug>` key can't be used without remap . I doubt anyone would be mad even if the internal `:map` mechanics treated `<Plug>` mappings to always be remap .